src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/CompareSymbolsCorrectlyTests.cs

```diff
@@ -538,5 +543,228 @@ class TestClass {{
+        {
+            await new VerifyCS.Test
+            {
+                TestState =
+                {
+                    Sources =
+                    {
+                        @"
+using System;
+using System.Collections.Immutable;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+public class C
+{
+    public void MethodWithDiagnostics(IEnumerable<KeyValuePair<ISymbol, int>> kvps, IEnumerable<ISymbol> symbols, ISymbol symbol)
+    {
+        [|ImmutableHashSet.Create<ISymbol>()|];
+        [|ImmutableHashSet.CreateBuilder<ISymbol>()|];
+        [|ImmutableHashSet.CreateRange(symbols)|];
+        [|symbols.ToImmutableHashSet()|];
+
+        [|ImmutableDictionary.Create<ISymbol, int>()|];
+        [|ImmutableDictionary.CreateBuilder<ISymbol, int>()|];
+        [|ImmutableDictionary.CreateRange(kvps)|];
+        [|kvps.ToImmutableDictionary()|];
+
+        [|symbols.Contains(symbol)|];
+        [|symbols.Distinct()|];
+        [|symbols.GroupBy(x => x)|];
+        [|symbols.GroupJoin(symbols, x => x, x => x, (x, y) => x)|];
+        [|symbols.Intersect(symbols)|];
+        [|symbols.Join(symbols, x => x, x => x, (x, y) => x)|];
+        [|symbols.SequenceEqual(symbols)|];
+        [|symbols.ToDictionary(x => x)|];
+        [|symbols.ToLookup(x => x)|];
+        [|symbols.Union(symbols)|];
+    }
+
+    public void MethodWithoutDiagnostics(IEnumerable<KeyValuePair<int, ISymbol>> kvps, IEnumerable<int> integers, int integer)
+    {
+        ImmutableHashSet.Create<int>();
+        ImmutableHashSet.CreateBuilder<int>();
+        ImmutableHashSet.CreateRange(integers);
+        integers.ToImmutableHashSet();
+
+        ImmutableDictionary.Create<int, ISymbol>();
+        ImmutableDictionary.CreateBuilder<int, ISymbol>();
+        ImmutableDictionary.CreateRange(kvps);
+        kvps.ToImmutableDictionary();
+
+        integers.Contains(integer);
+        integers.Distinct();
+        integers.GroupBy(x => x);
+        integers.GroupJoin(integers, x => x, x => x, (x, y) => x);
+        integers.Intersect(integers);
+        integers.Join(integers, x => x, x => x, (x, y) => x);
+        integers.SequenceEqual(integers);
+        integers.ToDictionary(x => x);
+        integers.ToLookup(x => x);
+        integers.Union(integers);
+    }
+}",
+                        SymbolEqualityComparerStubCSharp,
+                    },
+                },
+            }.RunAsync();
+
+            await new VerifyVB.Test
+            {
+                TestState =
+                {
+                    Sources =
+                    {
+                        @"
+Imports System
+Imports System.Collections.Immutable
+Imports System.Collections.Generic
+Imports System.Linq
+Imports Microsoft.CodeAnalysis
+
+Public Class C
+    Public Sub MethodWithDiagnostics(kvps As IEnumerable(Of KeyValuePair(Of ISymbol, Integer)), symbols As IEnumerable(Of ISymbol), symbol As ISymbol)
```

Comment:
For the cases where we would expect a diagnostic and where we don't have it (i.e. x4, x8, x9, x10, x13, x15 and x18), there is something weird with how the method is resolved.

For example, for [ToImmutableHashSet](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablehashset.toimmutablehashset), the docs considers the method as being generic while what we have is not a generic method:

![image](https://user-images.githubusercontent.com/11340282/87679129-1a56f500-c77c-11ea-9e99-fc6aeff7e3be.png)

Because of that we don't consider the method as acting on `ISymbol` so we bail-out.

Reply:
Ping @mavasani

Reply:
@Evangelink Can you look at `method.OriginalDefinition`?

Reply:
It doesn't work. I found a workaround but I cannot tell whether this is a bug to report or not.